sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-          #        return (dfa[state]["type"] == "1", stlst)
-
-      else:
-          try:
-              #print('weird place')
-              # take target id, discard counts
-              state = state[0]
-          except IndexError:
-              print("Out of alphabet: alternatives")
-          
-              stlst.append("-1")
-              if not statelist:
-                     return dfa[state]["type"] == "1"
-              else:
-                     return (dfa[state]["type"] == "1", stlst)
-      stlst.append(state)
-      if state in sinks and len(str(sev)) >= 2:
-          sev_sinks.add(state)
-  if not statelist:
-      return dfa[state]["type"] == "1"
-  else:
-      return (dfa[state]["type"] == "1", stlst, sev_sinks)
-
-def encode_sequences(m, m2):
-    #print(m2)
-    traces = []
-    sp = []
-    orig = []
-    with open(path_to_traces) as tf:
-        lines = tf.readlines()[1:]
-    #print(len(lines))
-
-    for line in lines:
-        if line == lines[-1]:
-             spl = line.split(' ')
-        else:
-            spl = line[:-1].split(' ')
-        
-        line = ' '.join(spl[2:])
-        #print(spl[2:])
-        orig.append([(x) for x in spl[2:] if x != ''])
-        traces.append(line)
-    num_sink = 0   
-    total = 0
+                    state = sinks[state_list[-1]][sym]
+                    state = '-1' if state == '' else state
+                except IndexError:
+                    state = '-1'  # With `printblue = 1` in spdfa-config.ini this should not happen
+            else:
+                state = '-1'
+
+        state_list.append(state)
+        if state in sinks and len(str(sev)) >= 2:
+            sev_sinks.add(state)
+
+    return state_list, sev_sinks
+
+
+# Step 6.2: Encode traces (include state IDs to mcat|mserv)
+def encode_sequences(dfa, sinks, flexfringe_traces):
+    traces_in_sinks, total_traces = 0, 0
     state_traces = dict()
-    for i,sample in enumerate(traces):
-        #print(sample)
-        r, s, _ = traverse(m, m2, sample, statelist=True)
-        s = [(x) for x in s]
-        sp.append(s)
-        state_traces[i] = s
-
-        total += len(s)
-        true = [1 if x == '-1' else 0 for x in s]
-        
-        num_sink += sum(true)
-        
-        #print('encoded', sample, state_traces[i])
-        assert (len(sample.split(' '))+1 == len(state_traces[i]))
+    med_sev_states, high_sev_states, sev_sinks = set(), set(), set()
+    for i, sample in enumerate(flexfringe_traces):
+        sample = ' '.join(sample.strip('\n').split(' ')[2:])  # Remove the first number and len(mcats)
+        state_list, _sev_sinks = traverse(dfa, sinks, sample)
+        state_traces[i] = state_list
 
-    #print(len(traces), len(state_traces))
-    print('Traces in sinks: ', num_sink, 'Total traces: ', total, 'Percentage: ',100*(num_sink/float(total)))
-    return (traces, state_traces)    
+        total_traces += len(state_list)  # TODO: decide what to do with the root
```

Comment:
Right now, the root node (`0`) is in `state_traces`. I think it should not be counted as trace, since it is artificial, but what do you think?

Reply:
The length of a subsequence is one less than the length of a state sequence (since [0] is added for an artificial transition). The new_state list actually gets rid of [0]. Maybe that format should be followed,

Reply:
But that happens later. Should I then do the cut `s[1:]` when the arrow is? 
![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/5ad57382-4d93-4e60-8018-2d328d55338d)

Reply:
[Updated as discussed in the meeting]